proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+  2030; they'll see just the warnings they need to upgrade to ``Stable2027``.
+
+* **Simplicity**. The goal is that a vast majority of our users will be
+  able to specify a language edition, and that's it. No extensions. No
+  warning flags. This simplifies what a user needs to think about when
+  setting up a Haskell project, removing the paralysis of choice that
+  can reign today.
+
+
+Proposed Change Specification
+-----------------------------
+
+-  We refresh the concept of *language
+   edition*. Existing language editions ``Haskell98``, ``Haskell2010``, and
+   ``GHC2021`` will continue to be language editions, to which we
+   add the following:
+
+   * ``Stable2024``
+   * ``Experimental2024``
+   * ``Latest``
+   * ``Student2024``
+
+- A language edition can be specified using the extensions syntax, by
+  passing e.g. ``-XStable2024`` on the command line or putting
+  ``Stable2024`` in a language pragma.
+
+- Every file is compiled with respect to precisely one language
+  edition. Specifying more than one edition is an error. If a user
+  specifies no edition during compilation, then a default is chosen:
+
+  * In all versions of GHC released before the end of 2025, the default
+    language edition will be ``GHC2021``.
+
+  * A new warning
+    ``-Wmissing-language-edition``, will inform users
+    that they should specify a language edition. This warning will be
+    off by default in GHC, but it is expected that cabal will turn it on.
+    In this way, someone experimenting with a standalone Haskell file doesn't
+    have to fuss about language editions, but anyone building a package
+    (presumably for publication) is encouraged to choose.
+
+  * Starting in 2026, the default language edition when none is specified
+    will be the latest ``Stable`` edition available. (The warning continues,
+    unchanged.)
+
+  * In GHCi, the default language edition will be the latest ``Stable``
+    edition. ``-Wmissing-language-edition`` will be off by default.
+
+- In GHCi, the top-level can have a distinct language edition from loaded
+  files. If a file is being loaded and does not specify its own language
+  edition, it inherits from the language edition in force at the top-level.
+
+- A language edition can control almost all behaviors of GHC. The meaning
+  (or existence) of other flags can depend on language edition. While
+  we will not implement it this way, we can imagine that GHC becomes
+  a set of programs that happen to share a binary; the choice of which
+  program is chosen by the language edition.
+
+  The one restriction on the expressive power of language editions
+  is that build products of different language editions (within the
+  same version of GHC) must be
+  compatible. We expect the Haskell ecosystem to contain packages
+  compiled with a variety of language editions, and they must work
+  together. The word *compatible* above is doing some heavy lifting,
```

Comment:
Yes, but if they're not compatible, then we've lost the baby with the bathwater. I agree that there is tension between the global and the local here, but I find this aspect still to be in balance. Do you see a place where this tension is causing trouble?

Reply:
I think this is maybe the wrong thread to have this discussion, since I think the way to think about this is whether the language edition is a global or local choice (the stuff about producing compatible code is downstream of that).

- Your suggestion that "GHC could become a set of binaries" is inconsistent with a local choice: then how do I compile a dependency with a different (local) language edition? My edition-X binary can compile edition-Y code (then nothing changes)? I need a bundle of binaries (ugh, complexity)? A set of binaries to me suggests a _global_ choice, where you have to make it for all your dependencies too.
- The "rock solid" user persona cannot be satisfied by a local version, since they want to enforce a global choice (stability) for all their dependencies.

I think there are more, but I really do think local/global is something we need to settle and which has consequences throughout.

Reply:
(My proposal https://github.com/ghc-proposals/ghc-proposals/pull/636#pullrequestreview-1903232732 proposes that language editions are local, but stability information is propagated so that users _can_ require stability globally.)